sys/dev/sound/pci/hda/hdaa_patches.c

```diff
@@ -366,63 +271,63 @@ hdac_pin_patch(struct hdaa_widget *w)
 	    subid == LENOVO_G580_SUBVENDOR)) {
 		switch (nid) {
 		case 25:
-			patch = "as=1 seq=15";
+			patch_str = "as=1 seq=15";
 			break;
 		/*
 		 * Group onboard mic and headphone mic
 		 * together.  Fixes onboard mic.
 		 */
 		case 27:
-			patch = "as=2 seq=15";
+			patch_str = "as=2 seq=15";
 			break;
 		case 35:
-			patch = "as=2";
-			break;
-		}
-	} else if (id == HDA_CODEC_ALC269 &&
-	    (subid == LENOVO_X1CRBN_SUBVENDOR ||
-	    subid == LENOVO_T430_SUBVENDOR ||
-	    subid == LENOVO_T430S_SUBVENDOR ||
-	    subid == LENOVO_T530_SUBVENDOR)) {
-		switch (nid) {
-		case 21:
-			patch = "as=1 seq=15";
-			break;
-		}
-	} else if (id == HDA_CODEC_ALC269 &&
-	    subid == ASUS_UX31A_SUBVENDOR) {
-		switch (nid) {
-		case 33:
-			patch = "as=1 seq=15";
+			patch_str = "as=2";
 			break;
 		}
-	} else if (id == HDA_CODEC_ALC892 &&
-	    subid == INTEL_DH87RL_SUBVENDOR) {
-		switch (nid) {
-		case 27:
-			patch = "as=1 seq=15";
-			break;
-		}
-	} else if (id == HDA_CODEC_ALC292 &&
-	    subid == LENOVO_X120BS_SUBVENDOR) {
-		switch (nid) {
-		case 21:
-			patch = "as=1 seq=15";
-			break;
+	} else {
+		/*
+		 * loop over hdaa_model_pin_patch
+		 */
+		struct pin_patch_t *pin_patches = NULL;
+
+		for (int ci = 0; ci < nitems(realtek_model_pin_patches); ci++) {
+			struct hdaa_model_pin_patch_t *p = &realtek_model_pin_patches[ci];
+			if (id == p->id) {
+				struct model_pin_patch_t *pinpatches = p->patches;
+				for (struct model_pin_patch_t *pp = pinpatches; pp->models; pp++) {
+					for (struct pin_machine_model_t *model = pp->models; model->id; model++) {
+						if (subid == model->id) {
+							pin_patches = pp->pin_patches;
+							break;
+						}
+					}
+				}
+			}
 		}
-	} else if (id == HDA_CODEC_ALC298 && subid == DELL_XPS9560_SUBVENDOR) {
-		switch (nid) {
-		case 24:
-			config  = 0x01a1913c;
-			break;
-		case 26:
-			config  = 0x01a1913d;
-			break;
+
+		for (struct pin_patch_t *patch = pin_patches; patch->type; patch++) {
+			if (nid == patch->nid) {
+				switch (patch->type) {
+				case PIN_PATCH_TYPE_STRING:
+					patch_str = patch->patch.string;
+				case PIN_PATCH_TYPE_MASK:
+					config &= ~patch->patch.mask[0];
+					config |= patch->patch.mask[1];
+					break;
+				case PIN_PATCH_TYPE_OVERRIDE:
+					config = patch->patch.override;
+					break;
+				default:
+					/* should panic hard */
+					break;
+				}
+				break;
+			}
 		}
 	}
 
-	if (patch != NULL)
-		config = hdaa_widget_pin_patch(config, patch);
+	if (patch_str)
```

Comment:
style(9): patch_str != NULL

Reply:
patch was in one of the git patches renamed to patch_str. patch is now a struct and not a string, and patch_str is initialized to NULL. if (patch_str) should therefore  be correct. Do you think otherwise?

Reply:
Somehow I missed the style(9) in your comment. Having reread style, I'll change this into a NULL check tomorrow.